File: golang/lib/services/container_creation_config.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package services

import "os"

const (
	defaultKurtosisVolumeMountpoint = "/kurtosis-enclave-data"
)

type StaticFileID string

// The ID of an artifact containing files that should be mounted into a service container
type FilesArtifactID string

// ====================================================================================================
//                                    Config Object
// ====================================================================================================
// TODO defensive copy when we're giving back complex objects?????
// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
type ContainerCreationConfig struct {
	image                    string
	kurtosisVolumeMountpoint string   // Technically the enclave data volume, but we call it this for simplicity for the user
	usedPortsSet             map[string]bool
	fileGeneratingFuncs      map[string]func(*os.File) error
	usedStaticFilesSet       map[StaticFileID]bool
	filesArtifactMountpoints map[FilesArtifactID]string
}

func (config *ContainerCreationConfig) GetImage() string {
	return config.image
}

func (config *ContainerCreationConfig) GetKurtosisVolumeMountpoint() string {
	return config.kurtosisVolumeMountpoint
}

func (config *ContainerCreationConfig) GetUsedPortsSet() map[string]bool {
	return config.usedPortsSet
}

func (config *ContainerCreationConfig) GetFileGeneratingFuncs() map[string]func(*os.File) error {
	return config.fileGeneratingFuncs
}

func (config *ContainerCreationConfig) GetFilesArtifactMountpoints() map[FilesArtifactID]string {
	return config.filesArtifactMountpoints
}

func (config *ContainerCreationConfig) GetUsedStaticFiles() map[StaticFileID]bool {
	return config.usedStaticFilesSet
}

// ====================================================================================================
//                                        Builder
// ====================================================================================================
// TODO Defensive copies on all these With... functions???
// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
type ContainerCreationConfigBuilder struct {
	image                    string
	kurtosisVolumeMountpoint string
	usedPortsSet             map[string]bool
	usedStaticFilesSet       map[StaticFileID]bool
	fileGeneratingFuncs      map[string]func(*os.File) error
	filesArtifactMountpoints map[FilesArtifactID]string
}

func NewContainerCreationConfigBuilder(image string) *ContainerCreationConfigBuilder {
	return &ContainerCreationConfigBuilder{
		image:                    image,
		kurtosisVolumeMountpoint: defaultKurtosisVolumeMountpoint,
		usedPortsSet:             map[string]bool{},
		fileGeneratingFuncs:      map[string]func(file *os.File) error{},
		filesArtifactMountpoints: map[FilesArtifactID]string{},
	}
}

func (builder *ContainerCreationConfigBuilder) WithKurtosisVolumeMountpoint(kurtosisVolumeMountpoint string) *ContainerCreationConfigBuilder {
	builder.kurtosisVolumeMountpoint = kurtosisVolumeMountpoint
	return builder
}

func (builder *ContainerCreationConfigBuilder) WithUsedPorts(usedPortsSet map[string]bool) *ContainerCreationConfigBuilder {
	builder.usedPortsSet = usedPortsSet
	return builder
}

func (builder *ContainerCreationConfigBuilder) WithGeneratedFiles(fileGeneratingFuncs map[string]func(*os.File) error) *ContainerCreationConfigBuilder {
	builder.fileGeneratingFuncs = fileGeneratingFuncs
	return builder
}

func (builder *ContainerCreationConfigBuilder) WithStaticFiles(usedStaticFilesSet map[StaticFileID]bool) *ContainerCreationConfigBuilder {
	builder.usedStaticFilesSet = usedStaticFilesSet
	return builder
}

func (builder *ContainerCreationConfigBuilder) WithFilesArtifacts(filesArtifactMountpoints map[FilesArtifactID]string) *ContainerCreationConfigBuilder {
	builder.filesArtifactMountpoints = filesArtifactMountpoints
	return builder
}

func (builder *ContainerCreationConfigBuilder) Build() *ContainerCreationConfig {
	return &ContainerCreationConfig{
		image:                    builder.image,
		kurtosisVolumeMountpoint: builder.kurtosisVolumeMountpoint,
		usedPortsSet:             builder.usedPortsSet,
		fileGeneratingFuncs:      builder.fileGeneratingFuncs,
		usedStaticFilesSet:       builder.usedStaticFilesSet,
		filesArtifactMountpoints: builder.filesArtifactMountpoints,
	}
}

File: typescript/src/lib/services/container_creation_config.ts
import { Result } from "neverthrow";

const DEFAULT_KURTOSIS_VOLUME_MOUNTPOINT: string = "/kurtosis-enclave-data";

export type StaticFileID = string;

// The ID of an artifact containing files that should be mounted into a service container
export type FilesArtifactID = string;

// ====================================================================================================
//                                    Config Object
// ====================================================================================================
// TODO defensive copy when we're giving back complex objects?????
// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
export class ContainerCreationConfig {
	
    private readonly image: string;
    private readonly kurtosisVolumeMountpoint: string; // Technically the enclave data volume, but we call it this for simplicity for the user
    private readonly usedPortsSet: Set<string>;
    private readonly fileGeneratingFuncs: Map<string, (fp: number) => Promise<Result<null, Error>>>; // File descriptors are just integers
    private readonly usedStaticFilesSet: Set<StaticFileID>;
    private readonly filesArtifactMountpoints: Map<FilesArtifactID, string>;

    constructor(
            image: string,
            kurtosisVolumeMountpoint: string,
            usedPortsSet: Set<string>,
            fileGeneratingFuncs: Map<string, (fp: number) => Promise<Result<null, Error>>>,
            usedStaticFilesSet: Set<StaticFileID>,
            filesArtifactMountpoints: Map<FilesArtifactID, string>){
        this.image = image;
        this.kurtosisVolumeMountpoint = kurtosisVolumeMountpoint;
        this.usedPortsSet = usedPortsSet;
        this.fileGeneratingFuncs = fileGeneratingFuncs;
        this.usedStaticFilesSet = usedStaticFilesSet;
        this.filesArtifactMountpoints = filesArtifactMountpoints;
    }

    public getImage(): string {
        return this.image;
    }

    public getKurtosisVolumeMountpoint(): string {
        return this.kurtosisVolumeMountpoint;
    }

    public getUsedPortsSet(): Set<string> {
        return this.usedPortsSet;
    }

    public getFileGeneratingFuncs(): Map<string, (fp: number) => Promise<Result<null, Error>>> {
        return this.fileGeneratingFuncs;
    }

    public getFilesArtifactMountpoints(): Map<FilesArtifactID, string> {
        return this.filesArtifactMountpoints;
    }

    public getUsedStaticFiles(): Set<StaticFileID> {
        return this.usedStaticFilesSet;
    }
}

// ====================================================================================================
//                                        Builder
// ====================================================================================================
// TODO Defensive copies on all these With... functions???
// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
export class ContainerCreationConfigBuilder {
    private readonly image: string;
    private kurtosisVolumeMountpoint: string;
    private usedPortsSet: Set<string>;
    private usedStaticFilesSet: Set<StaticFileID>;
    private fileGeneratingFuncs: Map<string, (fp: number) => Promise<Result<null, Error>>>;
    private filesArtifactMountpoints: Map<FilesArtifactID, string>;

    constructor (image: string) {
        this.image = image;
        this.kurtosisVolumeMountpoint = DEFAULT_KURTOSIS_VOLUME_MOUNTPOINT;
        this.usedPortsSet = new Set();
        this.usedStaticFilesSet = new Set();
        this.fileGeneratingFuncs = new Map();
        this.filesArtifactMountpoints = new Map();
    }

    public withKurtosisVolumeMountpoint(kurtosisVolumeMountpoint: string): ContainerCreationConfigBuilder {
        this.kurtosisVolumeMountpoint = kurtosisVolumeMountpoint;
        return this;
    }

    public withUsedPorts(usedPortsSet: Set<string>): ContainerCreationConfigBuilder {
        this.usedPortsSet = usedPortsSet;
        return this;
    }

    public withGeneratedFiles(fileGeneratingFuncs: Map<string, (fp: number) => Promise<Result<null, Error>>>): ContainerCreationConfigBuilder {
        this.fileGeneratingFuncs = fileGeneratingFuncs;
        return this;
    }

    public withStaticFiles(usedStaticFilesSet: Set<StaticFileID>): ContainerCreationConfigBuilder {
        this.usedStaticFilesSet = usedStaticFilesSet;
        return this;
    }

    public withFilesArtifacts(filesArtifactMountpoints: Map<FilesArtifactID, string>): ContainerCreationConfigBuilder {
        this.filesArtifactMountpoints = filesArtifactMountpoints;
        return this;
    }

    public build(): ContainerCreationConfig {
        return new ContainerCreationConfig(
            this.image,
            this.kurtosisVolumeMountpoint,
            this.usedPortsSet,
            this.fileGeneratingFuncs,
            this.usedStaticFilesSet,
            this.filesArtifactMountpoints,
        );
    }
}
